control-plane-operator/controllers/hostedcontrolplane/etcd/reconcile.go

```diff
@@ -1,209 +1,367 @@
+				PeerTransportSecurity: SecurityConfig{
+					CertFile:      "/etc/etcd/tls/peer/peer.crt",
+					KeyFile:       "/etc/etcd/tls/peer/peer.key",
+					CertAuth:      true,
+					TrustedCAFile: "/etc/etcd/tls/peer/peer-ca.crt",
+					AutoTLS:       false,
+				},
+				ClientTransportSecurity: SecurityConfig{
+					CertFile:      "/etc/etcd/tls/server/server.crt",
+					KeyFile:       "/etc/etcd/tls/server/server.key",
+					CertAuth:      true,
+					TrustedCAFile: "/etc/etcd/tls/server/server-ca.crt",
+					AutoTLS:       false,
+				},
 			},
-			Resources: []string{
-				"etcdclusters",
-				"etcdbackups",
-				"etcdrestores",
+			ASG: ASGConfig{
+				Provider: "sts",
 			},
-			Verbs: []string{
-				"*",
+			// TODO: Add snapshot support
+			Snapshot: SnapshotConfig{
+				Provider: "noop",
+				Interval: p.SnapshotInterval,
+				TTL:      p.SnapshotTTL,
 			},
 		},
+	}
+	configBytes, err := yaml.Marshal(ecoConfig)
+	if err != nil {
+		return err
+	}
+	cm.Data = map[string]string{
+		"config.yaml": string(configBytes),
+	}
+	return nil
+}
+
+func ReconcileDiscoveryService(service *corev1.Service, ownerRef config.OwnerRef) error {
+	if service.CreationTimestamp.IsZero() {
+		service.Spec.Type = corev1.ServiceTypeClusterIP
+		service.Spec.ClusterIP = corev1.ClusterIPNone
+	}
+	ownerRef.ApplyTo(service)
+
+	service.Spec.PublishNotReadyAddresses = true
+	service.Spec.Selector = etcdPodSelector
+	service.Spec.Ports = []corev1.ServicePort{
 		{
-			APIGroups: []string{
-				corev1.SchemeGroupVersion.Group,
-			},
-			Resources: []string{
-				"pods",
-				"services",
-				"endpoints",
-				"persistentvolumeclaims",
-				"events",
-			},
-			Verbs: []string{
-				"*",
-			},
+			Name:       "peer",
+			Protocol:   corev1.ProtocolTCP,
+			Port:       2380,
+			TargetPort: intstr.Parse("peer"),
 		},
 		{
-			APIGroups: []string{
-				appsv1.SchemeGroupVersion.Group,
-			},
-			Resources: []string{
-				"deployments",
-			},
-			Verbs: []string{
-				"*",
-			},
+			Name:       "etcd-client",
+			Protocol:   corev1.ProtocolTCP,
+			Port:       2379,
+			TargetPort: intstr.Parse("client"),
 		},
 		{
-			APIGroups: []string{
-				corev1.SchemeGroupVersion.Group,
-			},
-			Resources: []string{
-				"secrets",
-			},
-			Verbs: []string{
-				"get",
-			},
+			Name:       "http",
+			Protocol:   corev1.ProtocolTCP,
+			Port:       2378,
+			TargetPort: intstr.Parse("http"),
 		},
 	}
 	return nil
 }
 
-func ReconcileOperatorRoleBinding(roleBinding *rbacv1.RoleBinding, ownerRef config.OwnerRef) error {
-	ownerRef.ApplyTo(roleBinding)
-	serviceAccount := manifests.EtcdOperatorServiceAccount(roleBinding.Namespace)
-	roleBinding.RoleRef = rbacv1.RoleRef{
-		APIGroup: rbacv1.SchemeGroupVersion.Group,
-		Kind:     "Role",
-		Name:     "etcd-operator",
+func ReconcileClientService(service *corev1.Service, ownerRef config.OwnerRef) error {
+	if service.CreationTimestamp.IsZero() {
+		service.Spec.Type = corev1.ServiceTypeClusterIP
```

Comment:
While it won't actually work to change this it would be incorrect if this was ever different, so maybe remove the gating so this surfaces as error?